File: 2294/main.py
from sys import stdin
from math import inf

def solve(target: int, coins) -> int:
    # counts[x]: 현재까지 입력받은 동전들을 사용해, 'x원'을 만들고자 할 때
    # 필요한 최소 동전 수
    counts = [0] + [inf]*target

    # DP
    for coin in coins:
        for current in range(coin, target+1):
            counts[current] = min(counts[current], 1 + counts[current - coin])

    return counts[target]

def main():
    target = int(stdin.readline().split()[1])
    lines = stdin.read().split()
    answer = solve(target, map(int, lines))
    if answer == inf:
        print('-1')
    else:
        print(answer)

if __name__ == '__main__':
    main()

File: 2294/main.pas
// fpc main.pas -O2 -Coti -vwnh
type
  u32 = Cardinal;
var
  coins_count, target, i, coin, current, candidate, answer: u32;
  counts: array[0..10000] of u32;
begin
  read(coins_count, target);

  // counts[x]: 현재까지 입력받은 동전들을 사용해, 'x원'을 만들고자 할 때 필요한
  // 최소 동전 수. 불가능한 경우 10000 초과의 값이 저장됨.
  counts[0] := 0;
  for i := 1 to target do begin
    counts[i] := 10001;
  end;

  // Perform DP
  for i := 1 to coins_count do begin
    read(coin);
    for current := coin to target do begin
      candidate := 1 + counts[current - coin];
      if counts[current] > candidate then begin
        counts[current] := candidate
      end
    end
  end;

  answer := counts[target];
  if answer > 10000 then begin
    writeln('-1')
  end else begin
    writeln(answer)
  end
end.
